src/runtime-rs/crates/runtimes/virt_container/src/factory/mod.rs

```diff
@@ -0,0 +1,282 @@
+    let mut factory_config = FactoryConfig {
+        template: toml_config.factory.template,
+        template_path: toml_config.factory.template_path.clone(),
+        cache: toml_config.factory.vm_cache_number,
+        vm_cache: toml_config.factory.vm_cache_number > 0,
+        vm_config: VMConfig {
+            hypervisor_name: toml_config.runtime.hypervisor_name.clone(),
+            agent_name: toml_config.runtime.agent_name.clone(),
+            hypervisor_config: toml_config
+                .hypervisor
+                .get(&toml_config.runtime.hypervisor_name)
+                .cloned()
+                .unwrap_or_default(),
+            agent_config: toml_config
+                .agent
+                .get(&toml_config.runtime.agent_name)
+                .cloned()
+                .unwrap_or_default(),
+        },
+    };
+
+    // Template
+    if toml_config.factory.template {
+        if new_factory(&mut factory_config, toml_config, true)
+            .await
+            .is_ok()
+        {
+            info!(sl!(), "vm factory is on");
+        } else {
+            info!(sl!(), "vm factory is off");
+        }
+    } else {
+        info!(sl!(), "vm factory is not enabled");
+    }
+
+    Ok(())
+}
+
+pub async fn new_factory(
+    config: &mut FactoryConfig,
+    toml_config: TomlConfig,
+    fetch_only: bool,
+) -> Result<()> {
+    // Validate VMConfig
+    config.vm_config.valid().map_err(|e| {
+        error!(
+            sl!(),
+            "factory::new_factory(): VMConfig validate failed {:#?}", e
+        );
+        e
+    })?;
+
+    info!(sl!(), "factory::new_factory(): VMConfig validate ok");
+
+    // template mode
+    if !config.template {
+        error!(sl!(), "factory::new_factory(): template must be enabled");
+    } else {
+        if fetch_only {
+            info!(sl!(), "factory::new_factory(): template fetch");
+            // Construct PathBuf
+            let path: PathBuf = config.template_path.clone().into();
+            Template::fetch(config.vm_config.clone(), path)
+                .context("failed to fetch VM template")?;
+        } else {
+            info!(sl!(), "factory::new_factory(): template new");
+            // Construct PathBuf
+            let path: PathBuf = config.template_path.clone().into();
+            Template::new(config.vm_config.clone(), toml_config, path)
+                .await
+                .inspect_err(|e| error!(sl!(), "Failed to create new Template factory: {}", e))
+                .context("failed to initialize VM template factory")?;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn close_factory(config: &mut FactoryConfig) -> Result<()> {
+    let state_path = Path::new(&config.template_path);
+
+    let c_state_path = CString::new(
+        state_path
+            .to_str()
+            .context("template_path is not valid UTF-8")?,
+    )
+    .context("template_path contains null byte")?;
+
+    let result = unsafe { libc::umount(c_state_path.as_ptr()) };
```

Comment:
The nix crate (`nix::mount`) provides `umount`. Don't use `unsafe` unless we have to.